### DISCUSSION TITLE
Document WebSockets

### DISCUSSION THREAD

**Discussion by Kludex** [fastapi/fastapi#9015] [Questions] [answered]
### First check

* [X] I added a very descriptive title to this issue.
* [X] I used the GitHub search to find a similar issue and didn't find it.
* [X] I searched the FastAPI documentation, with the integrated search.
* [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [X] I already read and followed all the tutorial in the docs and didn't find an answer.
* [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [X] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

I want to be able to describe my WebSockets the same way as my HTTP endpoints. I know that OpenAPI 3.0 doesn't support it. But what if we integrate [asyncapi](https://www.asyncapi.com/)?

OpenAPI and AsyncAPI are not incompatible, they have a lot of similarities as the former was inspiration for the latter. Hence, it will not be difficult to add support for it. The idea would be to choose what goes to the `/docs` endpoint.

### Describe alternatives you've considered

Don't document WebSockets. 

### Additional context

There are a lot of issues asking a way to document WebSockets.

## Update
- 2021-04-14: [Linux Foundation takes control of AsyncAPI specifications](https://devops.com/linux-foundation-takes-control-of-asyncapi-specification/)

**Comment by stevenj** [comment]
What i do is use the websocket to create an internal rest API call on receive from the websocket.  And i can document those normally.  I inject the websocket identifiers as headers to the API call, so that the message docs look like a normal post, and i document the headers as being internal only and NOT coming from the websocket message.  The replies i document in the same way.  Asynchronous replies are a little challenging to document.  Perhaps the callback documentation method could be used for that (we haven't solved that yet.)

Is it perfect? No.
Does it work with what is already there? Yes.

AsyncAPI also has its issues.  Its really targeting Subscriber/Publisher type async which isn't the only kind and if you document a normal long lived async connection with it (that doesn't have Sub/Pub semantics), you still have to massage its style to represent it.  So far our devs are having no issue translating the OpenAPI docs of our Websocket messages into implementation.

**Comment by Faridalim** [comment]
@stevenj  

> websocket to create an internal rest API call on receive from the websocket.

Can you tell me the way to do it please? Do you proxy the websocket traffic as http to another fastapi instance? Or fastapi receive it as websocket and direct it to its own api endpoint?

**Comment by stevenj** [comment]
@Faridalim I create a "/websocket/receive" endpoint which i use internally, i call it using a unix socket which hypercorn is bound to (faster than tcp) I pass as credentials in a header the identity of the websocket (so that the API is clean) and i verify that the request has no ip-address and reject it if it does because a unix socket has no ip address, this prevents the endpoints being called externally.  Its pretty easy.

We also have a "/websocket/transmit" endpoint, which is completely artificial, we never call it from anything, but it contains all or TX models that we can send on the websocket and all the documentation we want to put about them.  Which we found is a good way to document the outgoing websocket messages (as a lot of them are not "replies" they are asynchronous)  It works good.

**Comment by espositocode** [comment]
@tiangolo mentioned in another [issue](https://github.com/tiangolo/fastapi/issues/1232) that websockets shouldn't be used very much, but I disagree. For sending notifications to the client, websockets make a lot of sense. Ideally, each websocket message would just be a serialized json of the pydantic schemas we already use. For example, if you have an `ObjectRead` schema being used at `/objects/{id}`, you may want to also use the same schema (or a subset of it) on your websocket channel when the Object changes.

The problem is, I don't know how to document my data and provide a standardized structure to my messages. I don't think the average user should necessarily contend with the wild west of websocket protocols when something like AsyncAPI exists and could be pulled in. It solves the same problems OpenAPI does but for websockets, and if people are using websockets they should benefit from the same solutions, IMHO.

Edit: I do understand wanting to avoid feature bloat for things like this. At the very least, it would be nice to see how AsyncIO would work with FastAPI. Maybe this works best as a 3rd party library that people can pull in if needed, with docs and examples.

**Comment by Faridalim** [comment]
I think the easiest way to do it is to receive websocket message and transform it into http request to another fastapi endpoint

**Comment by hasansezertasan** [comment]
Is there any progress or plan about this one?

**Reply by Kludex** [reply]
No.

**Comment by andrew222651** [comment, marked as answer]
Check out the [FastWS](https://github.com/endrekrohn/fastws) project

**Comment by mclate** [comment]
I've taken one more attempt to address this and wrote [FastAPI WebSocket Router](https://github.com/mclate/fastapi-ws-router)
This uses original FastAPI documentation and "fakes" WebSockets messages as regular http routes
